src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -127,522 +102,417 @@ cast n =
-                      ((i, j) : rest) -> let (z, w) = Haskell.last rest
-                                             (ris, rjs) = Haskell.unzip $ Haskell.init rest
-                                          in solveN (i, j) (ris, rjs) (z, w)
-
-                solve1 :: MonadBlueprint i a m => i -> i -> m [i]
-                solve1 i j = do
-                    z0 <- newAssigned (\v -> v i - v j + fromConstant (2 ^ registerSize @a @n @r :: Natural))
-                    (z, _) <- splitExpansion (highRegisterSize @a @n @r) 1 z0
-                    return [z]
-
-                solveN :: MonadBlueprint i a m => (i, i) -> ([i], [i]) -> (i, i) -> m [i]
-                solveN (i, j) (is, js) (i', j') = do
-                    s <- newAssigned (\v -> v i - v j + fromConstant (t + one))
-                    (k, b0) <- splitExpansion (registerSize @a @n @r) 1 s
-                    (zs, b) <- flip runStateT b0 $ traverse StateT (Haskell.zipWith fullSub is js)
-                    d <- newAssigned (\v -> v i' - v j')
-                    s'0 <- newAssigned (\v -> v d + v b + fromConstant t)
-                    (s', _) <- splitExpansion (highRegisterSize @a @n @r) 1 s'0
-                    return (k : zs <> [s'])
-
-                fullSub :: MonadBlueprint i a m => i -> i -> i -> m (i, i)
-                fullSub xk yk b = do
-                    d <- newAssigned (\v -> v xk - v yk)
-                    s <- newAssigned (\v -> v d + v b + fromConstant t)
-                    splitExpansion (registerSize @a @n @r) 1 s
-
-        negate (UInt x) = UInt $ circuitF (V.unsafeToVector <$> solve)
-            where
-                solve :: MonadBlueprint i a m => m [i]
-                solve = do
-                    j <- newAssigned (Haskell.const zero)
-
-                    xs <- V.fromVector <$> runCircuit x
-                    let y = 2 ^ registerSize @a @n @r
-                        ys = replicate (numberOfRegisters @a @n @r -! 2) (2 ^ registerSize @a @n @r -! 1)
-                        y' = 2 ^ highRegisterSize @a @n @r -! 1
-                        ns
-                            | numberOfRegisters @a @n @r Haskell.== 1 = [y' + 1]
-                            | otherwise = (y : ys) <> [y']
-                    (zs, _) <- flip runStateT j $ traverse StateT (Haskell.zipWith negateN ns xs)
-                    return zs
-
-                negateN :: MonadBlueprint i a m => Natural -> i -> i -> m (i, i)
-                negateN n i b = do
-                    r <- newAssigned (\v -> fromConstant n - v i + v b)
-                    splitExpansion (registerSize @a @n @r) 1 r
-
-
-instance (Arithmetic a, KnownNat n, KnownRegisterSize rs, r ~ NumberOfRegisters a n rs) => MultiplicativeSemigroup (UInt n rs (ArithmeticCircuit a)) where
-    UInt x * UInt y = UInt (circuitF $ V.unsafeToVector <$> solve)
+    UInt x - UInt y = UInt $ symbolic2F x y (\u v -> V.unsafeToVector $ V.fromVector u + V.fromVector v) solve
```

Comment:
Once again, pure implementation must come from the old instance with `Interpreter` context